### PR TITLE
Unified path structure

### PR DESCRIPTION
**Pull request by ivysrono**

### CODE REVIEW

File: src/setting.js [hunk 1]
```diff
@@ -5,7 +5,8 @@ const {app, BrowserWindow} = require('electron')
 
 for (let dir of [
   (exports.userDataDirectory =
-    process.env.PORTABLE_EXECUTABLE_DIR || app.getPath('userData')),
+    path.join(process.env.PORTABLE_EXECUTABLE_DIR, 'Sabaki') ||
```
Comment (yishn) [inline src/setting.js]: This won't work if `PORTABLE_EXECUTABLE_DIR` is undefined.

File: src/setting.js [hunk 2]
```diff
@@ -3,10 +3,11 @@ const fs = require('fs')
 const path = require('path')
 const {app, BrowserWindow} = require('electron')
 
+const portableDir = process.env.PORTABLE_EXECUTABLE_DIR
 for (let dir of [
-  (exports.userDataDirectory =
-    path.join(process.env.PORTABLE_EXECUTABLE_DIR, 'Sabaki') ||
-    app.getPath('userData')),
+  (exports.userDataDirectory = portableDir
+    ? path.join(process.env.PORTABLE_EXECUTABLE_DIR, 'Sabaki')
```
Comment (yishn) [inline src/setting.js]: You can use `portableDir` here as well.

**Comment by ivysrono** [comment]
emm, engines path need to resolve relative path to portable dir, where should I edit?

**Comment by yishn** [comment]
Engines path should resolve relatively to the executable path of the application.

**Comment by ivysrono** [comment]
Portable's exe path is temp path.

**Comment by yishn** [comment]
The current working directory is something that the operating system itself sets. We don't have any control over it.

**Comment by yishn** [comment]
The current working directory is something that the operating system itself sets. Maybe it can be changed by [`process.chdir()`](https://nodejs.org/api/process.html#process_process_chdir_directory)?

**Comment by ivysrono** [comment]
> 
> 
> Maybe it can be changed by [`process.chdir()`](https://nodejs.org/api/process.html#process_process_chdir_directory)?

I think so.

**Comment by ivysrono** [comment]
@yishn Why there is no portable in 0.51.0 release?

**Comment by yishn** [comment]
@ivysrono Since I thought you still wanted to add the `process.chdir()` stuff, I haven't released the portable version yet. I think the best place to add it would be in `App.js` right after `fixPath()`.

**Comment by ivysrono** [comment]
```
const process = require('process')
const portableDir = process.env.PORTABLE_EXECUTABLE_DIR
if (portableDir) {
  process.chdir(portableDir)
}
```
How to set relatively sub-folder path?
图片
图片

**Comment by yishn** [comment]
Try to see what kind of error message gets thrown in the developer tools.

**Comment by ivysrono** [comment]
```
events.js:187 Uncaught Error: spawn ./0/Leela0110GTP_OpenCL.exe ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:264)
    at onErrorNT (internal/child_process.js:456)
    at processTicksAndRejections (internal/process/task_queues.js:80)
errnoException @ internal/errors.js:474
ChildProcess._handle.onexit @ internal/child_process.js:264
onErrorNT @ internal/child_process.js:456
processTicksAndRejections @ internal/process/task_queues.js:80
```

**Pull request merged**